File: src/test/java/com/morak/performancetracker/aop/persistence/ConnectionAdviceTest.java
package com.morak.performancetracker.aop.persistence;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.BDDMockito.given;

import com.morak.performancetracker.context.Accumulator;
import com.morak.performancetracker.context.ResultMapper;
import org.aopalliance.intercept.MethodInvocation;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class ConnectionAdviceTest {

    @Mock
    private MethodInvocation methodInvocation;

    @Test
    void proxy된_객체가_반환되고_monitor에_query를_세팅한다() throws Throwable {
        // given
        given(methodInvocation.proceed()).willReturn(new Object());
        given(methodInvocation.getArguments()).willReturn(new Object[]{"insert into sql"});
        QueryMonitor queryMonitor = new QueryMonitor();
        ConnectionAdvice advice = new ConnectionAdvice(queryMonitor,
                new Accumulator(new ResultMapper()));

        // when
        Object result = advice.invoke(methodInvocation);

        // then
        Assertions.assertAll(
                () -> assertThat(result).isNotNull(),
                () -> assertThat(queryMonitor.getQuery()).isEqualTo("insert into sql")
        );
    }
}

File: src/test/kotlin/com/morak/performancetracker/aop/persistence/ConnectionAdviceTest.kt
package com.morak.performancetracker.aop.persistence

import com.morak.performancetracker.context.Accumulator
import com.morak.performancetracker.context.ResultMapper
import org.aopalliance.intercept.MethodInvocation
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.api.function.Executable
import org.mockito.BDDMockito
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
internal class ConnectionAdviceTest {
    @Mock
    private val methodInvocation: MethodInvocation? = null

    @Test
    @Throws(Throwable::class)
    fun proxy된_객체가_반환되고_monitor에_query를_세팅한다() {
        // given
        BDDMockito.given(methodInvocation!!.proceed()).willReturn(Any())
        BDDMockito.given(methodInvocation.arguments).willReturn(arrayOf<Any>("insert into sql"))
        val queryMonitor = QueryMonitor()
        val advice = ConnectionAdvice(
            queryMonitor,
            Accumulator(ResultMapper())
        )

        // when
        val result = advice.invoke(methodInvocation)

        // then
        org.junit.jupiter.api.Assertions.assertAll(
            Executable { assertThat(result).isNotNull() },
            Executable { assertThat(queryMonitor.query).isEqualTo("insert into sql") }
        )
    }
}
